okta_app_signon_policy_rules: "Provider produced inconsistent result after apply" on chains attribute due to JSON key ordering

### Community Note

* Please vote on this issue by adding a 👍 [reaction](https://blog.github.com/2016-03-10-add-reactions-to-pull-requests-issues-and-comments/) to the original issue to help the community and maintainers prioritize this request.
* Please do not leave _+1_ or _me too_ comments, they generate extra noise for issue followers and do not help prioritize the request.
* If you are interested in working on this issue or have submitted a pull request, please leave a comment.
        Before submitting a bug report, we ask that you first [search existing issues](https://github.com/okta/terraform-provider-okta/issues) and [pull requests](https://github.com/okta/terraform-provider-okta/pulls) to see if someone else may have experienced the same issue or may have already submitted a fix for it. This helps to keep all relevant information in one place, including any potential workarounds.

### A Note on Terraform Core Issues

We also ask that you consider whether your issue may be related to Terraform Core. If you are running into one of the following scenarios, we recommend [opening an issue](https://github.com/hashicorp/terraform/issues/new/choose) in the Terraform Core repository instead:

* [Configuration Language](https://developer.hashicorp.com/terraform/language) or resource ordering issues
* [State](https://developer.hashicorp.com/terraform/language/state) and [State Backend](https://developer.hashicorp.com/terraform/language/backend) issues
* [Provisioner](https://developer.hashicorp.com/terraform/language/resources/provisioners/syntax) issues
* [Registry](https://registry.terraform.io/) issues
* Issues that span resources across multiple providers


### Terraform Version & Okta Provider Version(s)

Terraform v1.14.9
on darwin_arm64
+ provider registry.terraform.io/okta/okta v6.10.0

### Affected Resource(s)

`okta_app_signon_policy_rules`

### Can this be done in the Admin UI?

Yes

### Can this be done in the actual API call?

Yes

### Customer Information

Organization Name: paddle
Paid Customer: yes


### Terraform Configuration

```tf
terraform {
  required_providers {
    okta = { source = "okta/okta", version = "~> 6.10.0" }
  }
}

resource "okta_app_signon_policy" "test" {
  name        = "chains-ordering-test"
  description = "Reproduces chains key ordering inconsistency"
}

resource "okta_app_signon_policy_rules" "test" {
  policy_id = okta_app_signon_policy.test.id

  rule {
    name     = "FastPass"
    priority = 1
    type     = "AUTH_METHOD_CHAIN"
    chains = [jsonencode({
      authenticationMethods = [{ key = "okta_password", method = "password" }]
      reauthenticateIn      = "PT43800H"
      next = [{
        authenticationMethods = [{
          hardwareProtection      = "REQUIRED"
          key                     = "okta_verify"
          method                  = "signed_nonce"
          phishingResistant       = "REQUIRED"
          userVerification        = "REQUIRED"
          userVerificationMethods = ["BIOMETRICS"]
        }]
        reauthenticateIn = "PT43800H"
      }]
    })]
  }
}
```


### Debug Output

Not sharing the full debug output; **please reconsider this requirement as the logs will contain raw HTTP requests, including access tokens.**

As expected, all HTTP calls succeed and the resources are created as expected, only this part being of interest:

```
026-05-13T15:08:41.452+0100 [DEBUG] provider.terraform-provider-okta_v6.10.0: Value switched to prior value due to semantic equality logic: @caller=github.com/hashicorp/terraform-plugin-framework@v1.18.0/internal/fwschemadata/value_semantic_equality.go:91 tf_mux_provider=tf6to5server.v6tov5Server tf_provider_addr=okta/okta @module=sdk.framework tf_attribute_path=policy_id tf_req_id=0a0bfe61-0e4f-bbbc-c5da-47c62ae8e538 tf_resource_type=okta_app_signon_policy_rules tf_rpc=ApplyResourceChange timestamp="2026-05-13T15:08:41.452+0100"
2026-05-13T15:08:41.452+0100 [DEBUG] provider.terraform-provider-okta_v6.10.0: Value switched to prior value due to semantic equality logic: tf_req_id=0a0bfe61-0e4f-bbbc-c5da-47c62ae8e538 tf_rpc=ApplyResourceChange @module=sdk.framework tf_attribute_path=rule[0].network_connection tf_mux_provider=tf6to5server.v6tov5Server tf_resource_type=okta_app_signon_policy_rules @caller=github.com/hashicorp/terraform-plugin-framework@v1.18.0/internal/fwschemadata/value_semantic_equality.go:91 tf_provider_addr=okta/okta timestamp="2026-05-13T15:08:41.452+0100"
2026-05-13T15:08:41.453+0100 [DEBUG] provider.terraform-provider-okta_v6.10.0: Value switched to prior value due to semantic equality logic: tf_mux_provider=tf6to5server.v6tov5Server tf_provider_addr=okta/okta tf_rpc=ApplyResourceChange @caller=github.com/hashicorp/terraform-plugin-framework@v1.18.0/internal/fwschemadata/value_semantic_equality.go:91 @module=sdk.framework tf_attribute_path=rule[0].type tf_req_id=0a0bfe61-0e4f-bbbc-c5da-47c62ae8e538 tf_resource_type=okta_app_signon_policy_rules timestamp="2026-05-13T15:08:41.452+0100"
2026-05-13T15:08:41.453+0100 [DEBUG] provider.terraform-provider-okta_v6.10.0: Value switched to prior value due to semantic equality logic: @module=sdk.framework tf_mux_provider=tf6to5server.v6tov5Server tf_resource_type=okta_app_signon_policy_rules @caller=github.com/hashicorp/terraform-plugin-framework@v1.18.0/internal/fwschemadata/value_semantic_equality.go:91 tf_attribute_path=rule[0].name tf_provider_addr=okta/okta tf_req_id=0a0bfe61-0e4f-bbbc-c5da-47c62ae8e538 tf_rpc=ApplyResourceChange timestamp="2026-05-13T15:08:41.452+0100"
2026-05-13T15:08:41.453+0100 [DEBUG] provider.terraform-provider-okta_v6.10.0: Value switched to prior value due to semantic equality logic: tf_attribute_path=rule[0].status tf_provider_addr=okta/okta tf_req_id=0a0bfe61-0e4f-bbbc-c5da-47c62ae8e538 tf_resource_type=okta_app_signon_policy_rules @module=sdk.framework tf_mux_provider=tf6to5server.v6tov5Server tf_rpc=ApplyResourceChange @caller=github.com/hashicorp/terraform-plugin-framework@v1.18.0/internal/fwschemadata/value_semantic_equality.go:91 timestamp="2026-05-13T15:08:41.452+0100"
2026-05-13T15:08:41.453+0100 [DEBUG] provider.terraform-provider-okta_v6.10.0: Value switched to prior value due to semantic equality logic: tf_provider_addr=okta/okta tf_rpc=ApplyResourceChange @caller=github.com/hashicorp/terraform-plugin-framework@v1.18.0/internal/fwschemadata/value_semantic_equality.go:91 tf_attribute_path=rule[0].priority tf_mux_provider=tf6to5server.v6tov5Server tf_req_id=0a0bfe61-0e4f-bbbc-c5da-47c62ae8e538 tf_resource_type=okta_app_signon_policy_rules @module=sdk.framework timestamp="2026-05-13T15:08:41.452+0100"
2026-05-13T15:08:41.453+0100 [DEBUG] provider.terraform-provider-okta_v6.10.0: Value switched to prior value due to semantic equality logic: @caller=github.com/hashicorp/terraform-plugin-framework@v1.18.0/internal/fwschemadata/value_semantic_equality.go:91 @module=sdk.framework tf_resource_type=okta_app_signon_policy_rules tf_attribute_path=rule[0].risk_score tf_mux_provider=tf6to5server.v6tov5Server tf_provider_addr=okta/okta tf_req_id=0a0bfe61-0e4f-bbbc-c5da-47c62ae8e538 tf_rpc=ApplyResourceChange timestamp="2026-05-13T15:08:41.452+0100"
2026-05-13T15:08:41.453+0100 [DEBUG] provider.terraform-provider-okta_v6.10.0: Value switched to prior value due to semantic equality logic: tf_provider_addr=okta/okta tf_resource_type=okta_app_signon_policy_rules tf_rpc=ApplyResourceChange @module=sdk.framework tf_attribute_path=rule[0].access tf_mux_provider=tf6to5server.v6tov5Server tf_req_id=0a0bfe61-0e4f-bbbc-c5da-47c62ae8e538 @caller=github.com/hashicorp/terraform-plugin-framework@v1.18.0/internal/fwschemadata/value_semantic_equality.go:91 timestamp="2026-05-13T15:08:41.452+0100"
2026-05-13T15:08:41.453+0100 [DEBUG] provider.terraform-provider-okta_v6.10.0: Value switched to prior value due to semantic equality logic: tf_req_id=0a0bfe61-0e4f-bbbc-c5da-47c62ae8e538 tf_resource_type=okta_app_signon_policy_rules @caller=github.com/hashicorp/terraform-plugin-framework@v1.18.0/internal/fwschemadata/value_semantic_equality.go:91 tf_attribute_path=rule[0].factor_mode tf_mux_provider=tf6to5server.v6tov5Server tf_rpc=ApplyResourceChange @module=sdk.framework tf_provider_addr=okta/okta timestamp="2026-05-13T15:08:41.452+0100"
2026-05-13T15:08:41.458+0100 [DEBUG] State storage *statemgr.Filesystem declined to persist a state snapshot
2026-05-13T15:08:41.459+0100 [ERROR] vertex "okta_app_signon_policy_rules.test" error: Provider produced inconsistent result after apply
```



### Expected Behavior

`terraform apply` succeeds when creating or replacing an `okta_app_signon_policy_rules` resource with `type = "AUTH_METHOD_CHAIN"` rules. Subsequent `terraform plan` shows no changes.

### Actual Behavior


`terraform apply` fails with "Provider produced inconsistent result after apply" on the `chains` attribute:

```
Error: Provider produced inconsistent result after apply

When applying changes to okta_app_signon_policy_rules.this,
provider "registry.terraform.io/okta/okta" produced an unexpected new value:
.rule[0].chains[0]:
  was cty.StringVal("{\"authenticationMethods\":[{\"hardwareProtection\":\"REQUIRED\",\"key\":\"okta_verify\",...}],...}")
  but now cty.StringVal("{\"authenticationMethods\":[{\"key\":\"okta_verify\",\"method\":\"signed_nonce\",\"hardwareProtection\":\"REQUIRED\",...}],...}")
```

The two strings represent the **same JSON object**; the only difference is key ordering within the `authenticationMethods` items. Because `chains` is a `list(string)`, Terraform compares them as raw strings and flags a mismatch.

### Steps to reproduce

Run `terraform apply` on the configuration shared above

### Important Factoids

The root cause is a mismatch between two JSON serialisation paths:
1. **user-provided input:** Terraform's `jsonencode` always sorts object keys alphabetically → `hardwareProtection` before `key`.

2. **provider read-back:** In [`resource_okta_app_signon_policy_rules.go` L1373–L1381](https://github.com/okta/terraform-provider-okta/blob/a9bf1d06be9f2e8b57e08ab092533dce83294cab/okta/services/idaas/resource_okta_app_signon_policy_rules.go#L1373-L1381) (and again at [L1560–L1568](https://github.com/okta/terraform-provider-okta/blob/a9bf1d06be9f2e8b57e08ab092533dce83294cab/okta/services/idaas/resource_okta_app_signon_policy_rules.go#L1560-L1568)), chains are read from the API and re-serialised with `json.Marshal(chain)` where `chain` is `*sdk.AccessPolicyChains`. Go's `encoding/json` marshals struct fields in declaration order, and in [`sdk/v2_verificationMethod.go` L19–L27](https://github.com/okta/terraform-provider-okta/blob/a9bf1d06be9f2e8b57e08ab092533dce83294cab/sdk/v2_verificationMethod.go#L19-L27) the `AuthenticationMethodAccessPolicy` struct declares `Key` and `Method` before `HardwareProtection`:

```go
type AuthenticationMethodAccessPolicy struct {
    Key                     string   `json:"key,omitempty"`    // declared first
    Method                  string   `json:"method,omitempty"` // second
    HardwareProtection      string   `json:"hardwareProtection,omitempty"`
    ...
}
```

This produces `key, method, hardwareProtection, ...` — not alphabetical — so the planned and returned strings never match.

**The singular resource (`okta_app_signon_policy_rule`) does not have this bug** because the SDKv2 resource applies [`StateFunc: utils.NormalizeDataJSON` and `DiffSuppressFunc: utils.NoChangeInObjectFromUnmarshaledJSON`](https://github.com/okta/terraform-provider-okta/blob/a9bf1d06be9f2e8b57e08ab092533dce83294cab/okta/services/idaas/resource_okta_app_signon_policy_rule.go#L229-L235) to the `chains` attribute. [`NormalizeDataJSON`](https://github.com/okta/terraform-provider-okta/blob/a9bf1d06be9f2e8b57e08ab092533dce83294cab/okta/utils/utils.go#L462-L470) round-trips through `map[string]interface{}`, which causes Go to sort all keys alphabetically on re-marshal — normalising both the user input and the API response to the same canonical form. The Framework resource has no equivalent.

### References

_No response_

--- a/okta/services/idaas/resource_okta_app_signon_policy_rules.go
+++ b/okta/services/idaas/resource_okta_app_signon_policy_rules.go
@@ -1370,12 +1370,12 @@ func (r *appSignOnPolicyRulesResource) updateRuleActionsFromAPI(ctx context.Cont
 		rule.InactivityPeriod = types.StringValue(vm.InactivityPeriod)
 	}
 
-	// Convert chains to JSON strings.
+	// Convert chains to JSON strings and normalize key order so it matches Terraform's jsonencode output.
 	if len(vm.Chains) > 0 {
 		var chainStrings []string
 		for _, chain := range vm.Chains {
 			if jsonBytes, err := json.Marshal(chain); err == nil {
-				chainStrings = append(chainStrings, string(jsonBytes))
+				chainStrings = append(chainStrings, utils.NormalizeDataJSON(string(jsonBytes)))
 			}
 		}
 		rule.Chains, _ = types.ListValueFrom(ctx, types.StringType, chainStrings)
@@ -1557,12 +1557,12 @@ func (r *appSignOnPolicyRulesResource) convertAPIActionsToModel(ctx context.Cont
 		rule.Constraints, _ = types.ListValueFrom(ctx, types.StringType, constraintStrings)
 	}
 
-	// Convert chains to JSON strings.
+	// Convert chains to JSON strings and normalize key order so it matches Terraform's jsonencode output.
 	if len(vm.Chains) > 0 {
 		var chainStrings []string
 		for _, chain := range vm.Chains {
 			if jsonBytes, err := json.Marshal(chain); err == nil {
-				chainStrings = append(chainStrings, string(jsonBytes))
+				chainStrings = append(chainStrings, utils.NormalizeDataJSON(string(jsonBytes)))
 			}
 		}
 		rule.Chains, _ = types.ListValueFrom(ctx, types.StringType, chainStrings)

